app/src/main/java/com/catalincretu/newsreader/feature/newslist/model/NewsListViewModel.java

```diff
@@ -0,0 +1,88 @@
+package com.catalincretu.newsreader.feature.newslist.model;
+
+import android.annotation.SuppressLint;
+import android.app.Application;
+
+import androidx.annotation.NonNull;
+import androidx.databinding.ObservableArrayList;
+import androidx.databinding.ObservableBoolean;
+import androidx.databinding.ObservableField;
+import androidx.databinding.ObservableList;
+import androidx.lifecycle.AndroidViewModel;
+
+import com.catalincretu.data.NewsRepository;
+import com.catalincretu.data.feature.model.Article;
+import com.catalincretu.newsreader.feature.newslist.listener.Handler;
+import com.catalincretu.newsreader.feature.newslist.reactive.SingleLiveEvent;
+
+import java.util.List;
+
+import io.reactivex.android.schedulers.AndroidSchedulers;
+
+
+public class NewsListViewModel extends AndroidViewModel implements Handler {
+
+
+    @NonNull
+    public final ObservableList<ArticleItemViewModel> articlesList;
+    private final static String LINK = "https://newsapi.org/";
+    private final NewsRepository repo;
+    public final ObservableBoolean visibility_views;
+    public final ObservableBoolean isLoading;
+    public final ObservableField<String> resultText;
+    public final SingleLiveEvent<Throwable> error;
+    public final SingleLiveEvent<String> openLink;
+
+
+    public NewsListViewModel(Application application, NewsRepository repo) {
+        super(application);
+        this.repo = repo;
+        this.articlesList = new ObservableArrayList<>();
+        this.isLoading = new ObservableBoolean();
+        this.visibility_views = new ObservableBoolean(true);
+        this.resultText = new ObservableField<>();
+        this.error = new SingleLiveEvent<>();
+        this.openLink = new SingleLiveEvent<>();
+
+    }
+
+    @SuppressLint("CheckResult")
+    public void refreshData() {
+        isLoading.set(true);
+        visibility_views.set(false);
+        repo.getNewsArticles()
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(
+                        this::onNewsArticlesReceived,
+                        this::onNewsArticlesError
+                );
+    }
+
+    private void onNewsArticlesReceived(@NonNull List<Article> articles) {
+        isLoading.set(false);
+
+        for (Article item : articles) {
```

Comment:
let's move this mapping inside the Rx flow, repo.getNewsArticles().map(...), and have onNewsArticlesReceived received this already mapped list of ArticleItemViewModel. Otherwise, adding one at a time this.articlesList.add(articleItemViewModel), can lead to several updates towards the views.